apollo-api/src/main/java/com/spotify/apollo/Request.java

```diff
@@ -67,14 +69,31 @@
 
   /**
    * The headers of the request message.
+   * Deprecated in favor of {@link Request#headerEntries()}
    */
+  @Deprecated
   Map<String, String> headers();
 
   /**
-   * A header of the request message, or empty if no header with that name is found.
+   * All the headers of the request message.
+   */
+  Iterable<Map.Entry<String, String>> headerEntries();
+
+
+  /**
+   * A header of the request message, looked up in a case insensitive way,
+   * or empty if no header with that name is found.
    */
   default Optional<String> header(String name) {
-    return Optional.ofNullable(headers().get(name));
+    Preconditions.checkArgument(name != null, "Header names cannot be null");
```

Comment:
Would it be better to permit nulls here, and reverse the comparison below so that it would just fail to find something, rather than throw an exception for a null? I'm not sure. Either way, if nulls shouldn't be permitted, then I think we should use `Objects.requireNonNull` for the argument check.